src/libsystemd-network/dhcp6-option.c

```diff
@@ -79,6 +79,52 @@ int dhcp6_option_append(uint8_t **buf, size_t *buflen, uint16_t code,
         return 0;
 }
 
+int dhcp6_option_append_vendor_option(uint8_t **buf, size_t *buflen, OrderedHashmap *vendor_options) {
+        uint32_t enterprise_identifier;
+        sd_dhcp6_option *options;
+        size_t total, offset;
+        Iterator i;
+        int r;
+
+        assert(buf);
+        assert(*buf);
+        assert(buflen);
+        assert(vendor_options);
+
+        ORDERED_HASHMAP_FOREACH(options, vendor_options, i) {
+                _cleanup_free_ uint8_t *p = NULL;
+                uint8_t *q;
+
+                enterprise_identifier = htobe32(SYSTEMD_PEN);
+
+                p = memdup(&enterprise_identifier, sizeof(enterprise_identifier));
+                if (!p)
+                        return -ENOMEM;
+
+                total = sizeof(enterprise_identifier);
+                offset = total;
+
+                q = realloc(p, total + 2 + 2 + options->length);
+                if (!q)
+                        return -ENOMEM;
```

Comment:
hmmm, i don't get this: why first allocate just the identifier, and then grow it for the option payload? why not allocate things with the right size-right away? i.e.:

```c
p = malloc(4 + 2 + 2 + options->length);
if (!p)
        return -ENOMEM;

unaligned_write_be32(p, SYSTEMD_PEN);
unaligned_write_be16(p + 4, options->option);
unaligned_write_be16(p + 6, options->length);
memcpy(p + 8, options->data, options->length);
```

or what am I missing?

Reply:
also, SYSTEMD_PEN should be configurable by the user, no? I mean, this is supposed to be useful to set arbitrary vendor options, i.e. not just of the systemd enterprise id?

Reply:
Yeah update. thanks !